### PR TITLE
Gutenboarding: add start and complete tracking events for the main flow

### PR DESCRIPTION
**Pull request by ramonjd**
## Changes proposed in this Pull Request

See  #41059

- adding methods for start and end
- adding a constant for the flow name (could also be used for API calls, e.g., sites/new)

## Testing instructions

From http://calypso.localhost:3000/new:

Create a site with a new account

Check that `calypso_signup_start` and `calypso_signup_complete` are called with the right params, e.g.,

Screen Shot 2020-04-17 at 11 57 11 am 

Now create another site, but this time add a `ref` param, e.g., http://calypso.localhost:3000/new?ref=hoochie

Check that `calypso_signup_start` and `calypso_signup_complete` are called with the right params, 
e.g.,

Screen Shot 2020-04-17 at 12 00 16 pm

### CODE REVIEW

**Comment by andrewserong** [comment]
I know this is in progress, but just quickly adding that I like the approach here @ramonjd! Looks good adding a utils module for abstracting the calls to `recordTracksEvent` so that we don't have to manually add in the gutenboarding flow everywhere, and adding the calls in a `useEffect` in the relevant components feels nice and straightforward (and clear from looking at the component what's being fired).

File: client/landing/gutenboarding/components/header/index.tsx [hunk 1]
```diff
@@ -182,6 +182,8 @@ const Header: FunctionComponent = () => {
 				return;
 			}
 			resetOnboardStore();
+			// TODO: remove this :)
+			console.log( 'Call `calypso_signup_complete` here?' );
```
Comment (andrewserong) [inline client/landing/gutenboarding/components/header/index.tsx:186]: That's where I was thinking we'd call it :)

**Comment by ramonjd** [comment]
> adding a utils module for abstracting the calls to recordTracksEvent so that we don't have to manually add in the gutenboarding flow everywhere, and adding the calls in a useEffect in the relevant

Thanks! Looks like https://github.com/Automattic/wp-calypso/pull/41167 goes a little further along these lines too.

File: client/landing/gutenboarding/components/header/index.tsx (hunk 1 above)
Comment (ramonjd) [reply client/landing/gutenboarding/components/header/index.tsx:186]: Was just testing the `/new` flow and noticed that `calypso_signup_complete` is called before checkout, namely because [all the steps in the prelaunch flow](https://github.com/Automattic/wp-calypso/blob/9793440/client/signup/config/flows-pure.js#L358) (all one of them) are satisfied **after** the plans step.  

Screen Shot 2020-04-16 at 7 45 10 pm

I guess that won't matter much so long as we use whatever flow id we use for `calypso_signup_complete` in Gutenboarding for our `calypso_signup_complete`  call.

**Comment by ramonjd** [comment]
Thanks for testing @andrewserong , and for fixing the dupe signup complete event. ❤️ 

Will test later today and 🚢

File: client/landing/gutenboarding/utils/analytics.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/**
```
Comment (roo2) [inline client/landing/gutenboarding/utils/analytics.ts:1]: I think that this file should live outside of utils because as far as I can see, `utils/` are only pure functions and the `recordTracksEvent` is a side effect

File: client/landing/gutenboarding/utils/analytics.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { FLOW_ID } from '../constants';
+
+/**
+ * Make tracks call with embedded flow.
+ *
+ * @param {string} eventId The name/id of the tracks event. Should be like `calypso_something_snake_case`
```
Comment (roo2) [inline client/landing/gutenboarding/utils/analytics.ts:14]: Nitpick:
```suggestion
 * @param {string} eventId The name/id of the tracks event. Must be in snake_case and prefixed with "calypso" e.g. `calypso_something_snake_case`
```

File: client/landing/gutenboarding/utils/analytics.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ */
+import { recordTracksEvent } from '@automattic/calypso-analytics';
+
+/**
+ * Internal dependencies
+ */
+import { FLOW_ID } from '../constants';
+
+/**
+ * Make tracks call with embedded flow.
```
Comment (roo2) [inline client/landing/gutenboarding/utils/analytics.ts:12]: nitpick: 
```suggestion
 * Record an event in tracks including an embedded signup flow id.
```

File: client/landing/gutenboarding/utils/analytics.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	if ( ! ref ) {
+		ref = new URLSearchParams( window.location.search ).get( 'ref' ) || ref;
+	}
+	trackEventWithFlow( 'calypso_signup_start', { ref } );
+}
+
+/**
+ * Analytics call at the completion  of a Gutenboarding flow
+ *
+ * @param {object} params A set of params to pass to analytics for signup completion
```
Comment (roo2) [inline client/landing/gutenboarding/utils/analytics.ts:40]: this jsdoc comment is out of date with the current params on the method

File: client/landing/gutenboarding/gutenboard.tsx [hunk 6]
```diff
@@ -61,6 +62,9 @@ export function Gutenboard() {
 			document.head.appendChild( linkBase );
 			document.head.appendChild( linkHeadings );
 		} );
+
+		// Track Gutenboarding start
```
Comment (roo2) [inline client/landing/gutenboarding/gutenboard.tsx:66]: probably don't need this comment I think

File: client/landing/gutenboarding/components/header/index.tsx [hunk 7]
```diff
@@ -184,10 +185,18 @@ const Header: FunctionComponent = () => {
 				go();
 				return;
 			}
+
+			recordOnboardingComplete( {
+				isNewSite: !! newSite,
+				isNewUser: !! newUser,
+			} );
+
+			setIsRedirecting( true );
```
Comment (roo2) [inline client/landing/gutenboarding/components/header/index.tsx:194]: the setIsRedirecting change is unrelated to analytics (I think) is it meant to go in?

Comment (andrewserong) [reply client/landing/gutenboarding/components/header/index.tsx:194]: @roo2 yep, I added it in to prevent the analytics event [from firing twice](https://github.com/Automattic/wp-calypso/pull/41129#pullrequestreview-395241574) (it's the same pattern as used above when someone adds in a domain).

File: client/landing/gutenboarding/utils/analytics.ts (hunk 3 above)
Comment (ramonjd) [reply client/landing/gutenboarding/utils/analytics.ts:14]: Excellent suggestion

File: client/landing/gutenboarding/utils/analytics.ts (hunk 2 above)
Comment (ramonjd) [reply client/landing/gutenboarding/utils/analytics.ts:1]: Hmm, that is a good point about side effects. Do you reckon it's worth moving it into a store with action/controller type setup (in another PR)? Or just create a new `analytics` dir for now?

File: client/landing/gutenboarding/utils/analytics.ts (hunk 5 above)
Comment (ramonjd) [reply client/landing/gutenboarding/utils/analytics.ts:40]: The object passed is `{ isNewUser = false, isNewSite = true } `.

Do you mean I should update the docs to include the params?

```
  * @param {object} params A set of params to pass to analytics for signup completion
  * @param {boolean} params.isNewUser Whether the user is newly signed up
  * @param {boolean} params.isNewSite Whether a new site is created in the flow
```

File: client/landing/gutenboarding/utils/analytics.ts (hunk 2 above)
Comment (roo2) [reply client/landing/gutenboarding/utils/analytics.ts:1]: hmm I was thinking just to put it at the top level alongside `path.ts`, as far as I can tell that seems to be where these kind of common files have been kept in other parts of the codebase.  Or I've seen a `lib` folder being used too like we have `client/lib/analytics/`. Or yea just an `analytics` dir from the top level would work too.

File: client/landing/gutenboarding/utils/analytics.ts (hunk 5 above)
Comment (roo2) [reply client/landing/gutenboarding/utils/analytics.ts:40]: ahh sorry I think I just misread the javascript syntax I read it as having two params `isNewUser`, and `isNewSite` 😅. Yep but if you added those two lines to the docs I think it would be an improvement

File: client/landing/gutenboarding/utils/analytics.ts (hunk 2 above)
Comment (roo2) [reply client/landing/gutenboarding/utils/analytics.ts:1]: yea I think best would be like `gutenboarding/lib/analytics/analytics.ts` and then later we could move any other shared code with side effects there like `path.ts`.

Comment (ramonjd) [reply client/landing/gutenboarding/utils/analytics.ts]: Ah cool. I'll move it into `gutenboarding/lib/analytics/index.ts` for now. You're right: we can worry about where to dump it later 🤣

**Comment by roo2** [comment]
Tested this and yep, it records the `ref` on the signup step start. there are definately some cases it doesn't handle like closing signup or clicking to login but those can be addressed in https://github.com/Automattic/wp-calypso/pull/41167
